File: src/main/js/formulas/logic/IFNA.ts
import { F7Exception } from "../../errors/F7Exception";
import { F7ExceptionName } from "../../errors/F7ExceptionName";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class IFNA extends AbstractFormula {
  static SELF: IFNA = new IFNA();
  NAME = FormulaName.IFNA;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 2, this.NAME);
    const value = this.collateralLookup(origin, values[0]);
    const first = Converters.first(value);
    if (
      first instanceof F7Exception &&
      Converters.castAsF7Exception(first).name === F7ExceptionName.NA
    ) {
      return this.collateralLookup(origin, values[1]);
    }
    return value;
  }
}

File: src/main/java/io/protobase/f7/formulas/logic/IFNA.java
package io.protobase.f7.formulas.logic;

import io.protobase.f7.errors.F7Exception;
import io.protobase.f7.errors.F7ExceptionName;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class IFNA extends AbstractFormula {
  public static FormulaName NAME = FormulaName.IFERROR;
  public static IFNA SELF = new IFNA();

  public IFNA() {
    super();
  }

  public IFNA(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 2, NAME);
    Object value = Converters.first(collateralLookup.apply(origin, values[0]));
    if ((value instanceof F7Exception && ((F7Exception) value).getName().equals(F7ExceptionName.NA))) {
      return Converters.first(collateralLookup.apply(origin, values[1]));
    }
    return value;
  }
}
